### PR TITLE
GPU Test Framework

### LINKED ISSUE #3622: Reconsider lavapipe
The wgpu sources have various comments here and there remarking on how lavapipe doesn't implement Vulkan well enough to fully support WebGPU, but I wonder if that's true any more. Apparently Linux Mesa lavapipe 22.1.2 [passes](https://www.khronos.org/conformance/adopters/conformant-products#submission_696) the Vulkan 1.3 CTS.

We should go through wgpu and see what works now and which caveats still apply. It would sure be nice to have lavapipe for software-based testing.

### LINKED ISSUE #3680: Hard to Tell if Test is Skipped
We have a rather complex set of situations where a test is either skipped or expected to fail. As such we can't use the standard `#[should_panic]` or `#[ignored]` as we need to make that decision at runtime based on the device we're running on.

This can cause problems where the developer is trying to fix a particular problem, runs the tests, sees that they all pass, not realizing that the test they're trying to measure against isn't actually running or has an expected fail on it.

We should find some way to express if something is skipped.

One possibility is to use [`libtest-mimic`](https://docs.rs/libtest-mimic/latest/libtest_mimic/) to allow us to add a "kind" to tests which is then visible to cargo-nextest or cargo test. We likely would need our own macro for tests in that case and would have interesting ramifications for wasm testing via wasm-pack.

### LINKED ISSUE #4097: Examples which exclude lavapipe wrongly exclude other drivers with mesa vendor id
**Description**
Examples and tests which exclude lavapipe use the vendor id to identify the device. Community drivers in Mesa use the same vendor id. To be more precise, any case where lavapipe is excluded should also consider the adapter name.

**Repro steps**
Well... I'm not going to encourage you to run Vulkan drivers (agxv) which are still in development.

**Expected vs observed behavior**
Tests and examples which skip/explicitly fail when using lavapipe should not cause other drivers with the mesa vendor id to fail

### PR DESCRIPTION
**Pull request by cwfitzgerald**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**

Closes https://github.com/gfx-rs/wgpu/issues/3680 
Closes #3622 
Closes https://github.com/gfx-rs/wgpu/issues/4097

**Description**

This revamps our testing framework in multiple different ways.
- We now automatically test against all adapters in a given system.
- Supports async and sync code in our tests.
- Improves the visibility to skipped tests.

**How to Review**

While there are a ton of changes in this MR, a lot of it is very boring structural changes in the code. Please focus on `tests/src` `examples/common`, `wgpu-macros` and `xtask`.

**Testing**

It is lol

### CODE REVIEW

**Comment by cwfitzgerald** [comment]
Hey me! Rebase this!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ env:
   CARGO_TERM_COLOR: always
   RUST_LOG: info
   RUST_BACKTRACE: full
-  MSRV: 1.65
+  MSRV: 1.70
```
Comment (cwfitzgerald) [inline .github/workflows/ci.yml]: Figure this out

File: wgpu-macros/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+            ::wgpu_test::infra::TEST_LIST.lock().push(
+                // Allow any type that can be converted to a GpuTestConfiguration
+                ::wgpu_test::infra::GpuTestConfiguration::from(#expr).name_from_init_function_typename::<S>(#ident_lower)
+            )
+        }
+
+        #[cfg(target_arch = "wasm32")]
+        fn #test_name_webgl() {
+            // todo
```
Comment (cwfitzgerald) [inline wgpu-macros/src/lib.rs]: Figure this out

File: tests/tests/shader/zero_init_workgroup_mem.rs [hunk 3]
```diff
@@ -27,151 +26,153 @@ fn zero_init_workgroup_mem() {
...
-    MAX_DEVICE_WORKGROUP_MEMORY / (SHADER_WORKGROUP_MEMORY * TOTAL_WORK_GROUPS) + 1; // TODO: use div_ceil once stabilized
+            ))
+            .skip(FailureCase::backend_adapter(
+                Backends::VULKAN,
+                "llvmpipe",
+            ))
+            .limits(Limits::downlevel_defaults()),
+    )
+    .run_sync(|ctx| {
+        let bgl = ctx
```
Comment (cwfitzgerald) [inline tests/tests/shader/zero_init_workgroup_mem.rs:29]: These changes look suspisious

File: tests/tests/device.rs [hunk 4]
```diff
@@ -1,45 +1,32 @@
...
+                layout: &bind_group_layout,
+                entries: &[],
+            });
+        }
+
+        ctx.device.poll(wgpu::Maintain::Poll);
+    });
 
 #[cfg(not(all(target_arch = "wasm32", not(target_os = "emscripten"))))]
 #[test]
```
Comment (cwfitzgerald) [inline tests/tests/device.rs:30]: This needs to be wired up.

**Comment by nical** [comment]
It looks like metal is not tested anymore.

**Comment by nical** [comment]
> It looks like metal is not tested anymore.

Nevermind that.

**Comment by cwfitzgerald** [comment]
Alright, I think this MR is finally good to land whenever anyone can find time to review it - thankfully it's not actually as large as it seems:

`git diff origin/trunk --stat tests/src examples/common wgpu-macros xtask`

is

`16 files changed, 1280 insertions(+), 782 deletions(-)`

**Review by teoxoy — APPROVED** [review]
Took a brief look, nothing looks off.

The new features look great!
Even automatic retry and reporting of flaky tests! 🎉 

Shippit! 🚀

**Comment by cwfitzgerald** [comment]
WOOOOOOOOO

**Pull request merged**